Determine autograd engine ready queue based on InputMetadata instead of InputBuffer

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135633

Thanks @awgu for raising this issue and the small repro

From offline discussion with @albanD, in the case where a forward returns multiple outputs with different devices, we'd want to select the ready queue based on the device of the first one. Even though this is somewhat arbitrary, we prefer this over deciding which ready queue to push based on whichever input buffer's we happen to compute last, which can vary depending on more factors and thus be harder to reason about. This is in theory bc-breaking, but it seems unlikely that someone would depend on this behavior.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

soulitzer committed:
Determine autograd engine reqdyqueue based on first output's device

[ghstack-poisoned]

awgu commented on test/test_autograd.py:
```diff
@@ -4332,6 +4332,52 @@ def run_test(input_size, exponent):
+            executed.append("B")
+
+        x = torch.randn(
+            (1024, 1024), dtype=torch.bfloat16, device="cuda", requires_grad=True
```
haha I think we can make these a bit smaller for CI

soulitzer committed:
Update on "Determine autograd engine reqdyqueue based on first output's device"


Thanks awgu for raising this issue and the small repro

From offline discussion with albanD, in the case where a forward returns multiple outputs with different devices, we'd want to select the ready queue based on the device of the first one. Even though this is somewhat arbitrary, we prefer this over deciding which ready queue to push based on whichever input buffer's we happen to compute last, which can vary depending on more factors and thus be harder to reason about. This is in theory bc-breaking, but it seems unlikely that someone would depend on this behavior.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1107,7 +1107,8 @@ void Engine::evaluate_function(
       if (is_ready) {
-        auto queue = ready_queue(cpu_ready_queue, input_buffer.device());
+        // NB: The first forward output's device determines the queue
+        auto queue = ready_queue(cpu_ready_queue, next.function->input_metadata(0).device());
```
I think we want to preserve the non-CPU device being prioritized.
Could you update this by passing the input metadatas to the InputBuffer struct and make the .device() method there use that instead of the current buffer?

soulitzer replied:
Ah good catch!

soulitzer committed:
Update on "Determine autograd engine ready queue based on first output's device"


Thanks awgu for raising this issue and the small repro

From offline discussion with albanD, in the case where a forward returns multiple outputs with different devices, we'd want to select the ready queue based on the device of the first one. Even though this is somewhat arbitrary, we prefer this over deciding which ready queue to push based on whichever input buffer's we happen to compute last, which can vary depending on more factors and thus be harder to reason about. This is in theory bc-breaking, but it seems unlikely that someone would depend on this behavior.

[ghstack-poisoned]

soulitzer committed:
Update on "Determine autograd engine ready queue based on first output's device"


Thanks awgu for raising this issue and the small repro

From offline discussion with albanD, in the case where a forward returns multiple outputs with different devices, we'd want to select the ready queue based on the device of the first one. Even though this is somewhat arbitrary, we prefer this over deciding which ready queue to push based on whichever input buffer's we happen to compute last, which can vary depending on more factors and thus be harder to reason about. This is in theory bc-breaking, but it seems unlikely that someone would depend on this behavior.

[ghstack-poisoned]

soulitzer committed:
Update on "Determine autograd engine ready queue based on InputMetadata instead of InputBuffer"


Thanks awgu for raising this issue and the small repro

From offline discussion with albanD, in the case where a forward returns multiple outputs with different devices, we'd want to select the ready queue based on the device of the first one. Even though this is somewhat arbitrary, we prefer this over deciding which ready queue to push based on whichever input buffer's we happen to compute last, which can vary depending on more factors and thus be harder to reason about. This is in theory bc-breaking, but it seems unlikely that someone would depend on this behavior.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM !

awgu commented:
thanks for this @soulitzer !

soulitzer commented:
@pytorchbot merge
